src/query/functions/linear/sort.go

```diff
@@ -20,13 +20,178 @@
 
 package linear
 
-const (
-	// NB: Because Prometheus's sort and sort_desc only look at the last value,
-	// these functions are essentially noops in M3 as we don't support instant queries.
+import (
+	"fmt"
+	"sort"
+
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+)
 
+const (
 	// SortType returns timeseries elements sorted by their values, in ascending order.
 	SortType = "sort"
 
 	// SortDescType is the same as sort, but sorts in descending order.
 	SortDescType = "sort_desc"
 )
+
+type sortOp struct {
+	opType string
+	lessFn lessFn
+}
+
+// OpType for the operator
+func (o sortOp) OpType() string {
+	return o.opType
+}
+
+// String representation
+func (o sortOp) String() string {
+	return fmt.Sprintf("type: %s", o.opType)
+}
+
+type sortNode struct {
+	op         sortOp
+	controller *transform.Controller
+}
+
+type valueAndMeta struct {
+	val        float64
+	seriesMeta block.SeriesMeta
+}
+
+type lessFn func (i, j float64) bool
+
+// Node creates an execution node
+func (o sortOp) Node(
+	controller *transform.Controller,
+	_ transform.Options,
+) transform.OpNode {
+	return &sortNode{
+		op:         o,
+		controller: controller,
+	}
+}
+
+func (n *sortNode) Params() parser.Params {
+	return n.op
+}
+
+func (n *sortNode) Process(queryCtx *models.QueryContext, ID parser.NodeID, b block.Block) error {
+	return transform.ProcessSimpleBlock(n, n.controller, queryCtx, ID, b)
+}
+
+func (n *sortNode) ProcessBlock(queryCtx *models.QueryContext, ID parser.NodeID, b block.Block) (block.Block, error) {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	instantaneous := queryCtx.Options.Instantaneous
+	meta := b.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+
+	if instantaneous {
+		return n.processInstantBlock(queryCtx, stepIter, meta, seriesMetas)
+	}
+	return n.processNopBlock(queryCtx, stepIter, meta, seriesMetas)
+}
+
+func (n *sortNode) processInstantBlock(queryCtx *models.QueryContext, stepIter block.StepIter, meta block.Metadata, seriesMetas []block.SeriesMeta) (block.Block, error) {
+	for index := 0; stepIter.Next(); index++ {
+		if isLastStep(index, stepIter.StepCount()) {
+			meta.ResultMetadata.KeepNans = block.BoolTrue
+			//we only care for the last step values for the instant query
+			values := stepIter.Current().Values()
+			valuesToSort := make([]valueAndMeta, len(values))
```

Comment:
nit: using append is generally safer than indexed inserts, and doesn't make any real difference in performance

Reply:
Not sure how current code is unsafe? `append` needs to check current capacity and grow when needed which we do not want here (more possible allocations) because we already know the actual length and capacity for our value arrays.